src/adapter/graphql/src/queries/datasets/adapters/versioned_file.rs

```diff
@@ -0,0 +1,180 @@
+                )
+                .await
+                .map(|res| domain::GetDataResponse {
+                    df: res
+                        .df
+                        .map(|df| df.filter(col("version").eq(lit(version))).unwrap()),
+                    dataset_handle: res.dataset_handle,
+                    block_hash: res.block_hash,
+                })
+        } else {
+            query_svc
+                .tail(
+                    &self.state.dataset_handle().as_local_ref(),
+                    0,
+                    1,
+                    domain::GetDataOptions::default(),
+                )
+                .await
+        }
+        .int_err()?;
+
+        let Some(df) = query_res.df else {
+            return Ok(None);
+        };
+
+        let records = df.collect_json_aos().await.int_err()?;
+        if records.is_empty() {
+            return Ok(None);
+        }
+
+        assert_eq!(records.len(), 1);
+        let entry = VersionedFileEntry::from_json(
+            self.state.resolved_dataset(ctx).await?.clone(),
+            records.into_iter().next().unwrap(),
+        );
+        Ok(Some(entry))
+    }
+}
+
+#[common_macros::method_names_consts(const_value_prefix = "GQL: ")]
+#[Object]
+impl VersionedFile {
+    const DEFAULT_VERSIONS_PER_PAGE: usize = 100;
+
+    /// Returns list of versions in reverse chronological order
+    #[tracing::instrument(level = "info", name = VersionedFile_versions, skip_all)]
+    pub async fn versions(
+        &self,
+        ctx: &Context<'_>,
+        #[graphql(desc = "Upper version bound (inclusive)")] max_version: Option<FileVersion>,
+        page: Option<usize>,
+        per_page: Option<usize>,
+    ) -> Result<VersionedFileEntryConnection> {
+        use datafusion::logical_expr::{col, lit};
+
+        let page = page.unwrap_or(0);
+        let per_page = per_page.unwrap_or(Self::DEFAULT_VERSIONS_PER_PAGE);
+
+        let query_svc = from_catalog_n!(ctx, dyn domain::QueryService);
+
+        let query_res = query_svc
+            .get_data(
+                &self.state.dataset_handle().as_local_ref(),
+                domain::GetDataOptions::default(),
+            )
+            .await
+            .int_err()?;
+
+        let Some(df) = query_res.df else {
+            return Ok(VersionedFileEntryConnection::new(
+                Vec::new(),
+                page,
+                per_page,
+                0,
+            ));
+        };
+
+        let df = if let Some(max_version) = max_version {
+            df.filter(col("version").lt_eq(lit(max_version)))
+                .int_err()?
+        } else {
+            df
+        };
+
+        let total_count = df.clone().count().await.int_err()?;
+
+        let df = df
+            .sort(vec![col("version").sort(false, false)])
+            .int_err()?
+            .limit(page * per_page, Some(per_page))
+            .int_err()?;
+
+        let records = df.collect_json_aos().await.int_err()?;
```

Comment:
Q: do we need to take the `total_count` based on `records`?

Reply:
we calculate `total_count` before sorting and limiting the `df` to fetch current page, so I think it's correct - `total_count` should reflect how many versions there are in total, not just on the current page